File: travesim-main/controllers/referee_controller/referee_controller.cpp
/**
 * @file referee_controller.cpp
 *
 * @author Felipe Gomes de Melo <felipegmelo.42@gmail.com>
 *
 * @brief
 *
 * @date 07/2025
 *
 * @copyright MIT License - Copyright (c) 2025 Futebol Mini
 */

#include <ostream>
#include <webots/Robot.hpp>
#include <webots/Supervisor.hpp>
#include <webots/Node.hpp>
#include <webots/Emitter.hpp>
#include <memory>
#include <iostream>
#include <unordered_map>
#include <string>
#include <functional>

#include "travesim_adapters/data/data_common.hpp"
#include "travesim_webots/robot.hpp"

#include "travesim_webots/message.hpp"

#include "travesim_adapters/protobuf/replacer_receiver.hpp"
#include "travesim_adapters/protobuf/team_receiver.hpp"
#include "travesim_adapters/protobuf/vision_sender.hpp"
#include "travesim_adapters/data/field_state.hpp"

#define DEFAULT_Z_M 0.002 // 2mm

inline std::shared_ptr<std::unordered_map<std::string, webots::Node*>> get_robots_references(webots::Field* children) {
    auto entities = std::make_shared<std::unordered_map<std::string, webots::Node*>>();
    auto count = children->getCount();

    for (int i = 0; i < count; i++) {
        webots::Node* node = children->getMFNode(i);
        auto* name = node->getField("name");

        if (name) {
            (*entities)[name->getSFString()] = node;
        }
    }

    return entities;
}

inline void convert_to_entity_state(travesim::EntityState& output, travesim::webots_adapter::Robot& input) {
    output.position = input.get_position2d();
    output.angular_position = input.get_yaw();
    output.velocity = input.get_linear_velocity();
    output.angular_velocity = input.get_angular_velocity();
}

inline std::string build_name_from_team_number(bool is_yellow, uint8_t number) {
    return std::string(is_yellow ? "Yellow" : "Blue") + std::string("Robot") + std::to_string(number);
}

int main(int argc, char** argv) {
    /**
     * External interfaces definitions
     */

    const size_t robots_per_team = std::stoi(argv[1]);

    std::string referee_address_str(argv[2]);
    uint32_t referee_port = std::stoi(argv[3]);

    std::string yellow_address_str(argv[4]);
    uint32_t yellow_port = std::stoi(argv[5]);

    std::string blue_address_str(argv[6]);
    uint32_t blue_port = std::stoi(argv[7]);

    std::string multicast_addr_str(argv[8]);
    uint32_t multicast_port = std::stoi(argv[9]);

    bool specific_source = false;
    const travesim::TeamsFormation teams_formation = std::invoke([robots_per_team] {
        switch (robots_per_team) {
            case 3:
                return travesim::THREE_ROBOTS_PER_TEAM;

            case 5:
                return travesim::FIVE_ROBOTS_PER_TEAM;

            default:
                std::cout << "Invalid robots_per_team value! Got " << robots_per_team << " should be 3 or 5" << std::endl;
                std::exit(-1);
        }
    });

    std::cout << "Referee addr: " << referee_address_str << std::endl;
    std::cout << "Referee port: " << referee_port << std::endl;

    std::cout << "Yellow addr: " << yellow_address_str << std::endl;
    std::cout << "Yellow port: " << yellow_port << std::endl;

    std::cout << "Blue addr: " << blue_address_str << std::endl;
    std::cout << "Blue port: " << blue_port << std::endl;

    std::cout << "Vision addr: " << multicast_addr_str << std::endl;
    std::cout << "Vision port: " << multicast_port << std::endl;

    travesim::proto::VisionSender vision_sender(multicast_addr_str, multicast_port);
    travesim::FieldState field_state(teams_formation);

    travesim::proto::TeamReceiver yellow_receiver(yellow_address_str, yellow_port, true, specific_source, teams_formation);

    travesim::TeamCommand yellow_command(teams_formation);

    travesim::proto::TeamReceiver blue_receiver(blue_address_str, blue_port, false, specific_source, teams_formation);

    travesim::TeamCommand blue_command(teams_formation);

    travesim::proto::ReplacerReceiver referee_receiver(referee_address_str, referee_port, specific_source);
    std::queue<std::shared_ptr<travesim::EntityState>> states_queue;

    /**
     * Webots interfaces definitions
     */

    auto referee = std::make_unique<webots::Supervisor>();

    auto time_step = (uint16_t) referee->getBasicTimeStep();

    webots::Field* children = referee->getRoot()->getField("children");

    auto robots = get_robots_references(children);

    webots::Emitter* yellow_team_emitter = referee->getEmitter("yellow_team");
    webots::Emitter* blue_team_emitter = referee->getEmitter("blue_team");

    yellow_team_emitter->setChannel(0);
    blue_team_emitter->setChannel(1);

    /**
     * Loop preparation
     */

    uint32_t frame = 0;

    std::vector<travesim::webots_adapter::Robot> yellow_robots;
    yellow_robots.reserve(robots_per_team);

    std::vector<travesim::webots_adapter::Robot> blue_robots;
    blue_robots.reserve(robots_per_team);

    travesim::webots_adapter::Robot ball((*robots)["VssBall"]);

    for (size_t i = 0; i < robots_per_team; i++) {
        std::string yellow_robot_name = "YellowRobot" + std::to_string(i);
        std::string blue_robot_name = "BlueRobot" + std::to_string(i);

        yellow_robots[i] = travesim::webots_adapter::Robot((*robots)[yellow_robot_name]);
        blue_robots[i] = travesim::webots_adapter::Robot((*robots)[blue_robot_name]);
    }

    while (referee->step(time_step) != -1) {
        /**
         * Process messages from referee
         */

        bool received_new_msg = referee_receiver.receive(&states_queue);

        if (received_new_msg) {
            referee->simulationSetMode(webots::Supervisor::SIMULATION_MODE_PAUSE);

            while (!states_queue.empty()) {
                if (std::dynamic_pointer_cast<travesim::RobotState>(states_queue.front()) != nullptr) {
                    // Incoming state is a robot state
                    auto state = std::dynamic_pointer_cast<travesim::RobotState>(states_queue.front());

                    auto& team_array = (state->is_yellow ? yellow_robots : blue_robots);

                    team_array[state->id].set_position(state->position.x, state->position.y, DEFAULT_Z_M);
                    team_array[state->id].set_yaw(state->angular_position);
                    team_array[state->id].stop();
                } else {
                    // Incoming state is an entity state (ball)
                    auto state = std::dynamic_pointer_cast<travesim::EntityState>(states_queue.front());

                    ball.set_position(state->position.x, state->position.y, DEFAULT_Z_M);
                    ball.set_yaw(state->angular_position);
                    ball.stop();
                }

                states_queue.pop();
            }

            referee->simulationSetMode(webots::Supervisor::SIMULATION_MODE_REAL_TIME);
        }

        /**
         * Send world info to teams
         */

        // FIXME: Use std::chronos
        // Time comes in seconds, but time_step is in milliseconds
        field_state.time_step = (uint16_t) (referee->getTime() * 1e3 / time_step);

        for (size_t i = 0; i < robots_per_team; i++) {
            convert_to_entity_state(field_state.yellow_team[i], yellow_robots[i]);
            convert_to_entity_state(field_state.blue_team[i], blue_robots[i]);
        }

        convert_to_entity_state(field_state.ball, ball);

        vision_sender.send(&field_state);

        /**
         * Wait for teams to send a command
         */

        yellow_receiver.receive(&yellow_command);
        blue_receiver.receive(&blue_command);

        /**
         * Relay command to robots
         */

        frame++;

        travesim::webots_adapter::message_t<MAX_ROBOTS> yellow_message;
        travesim::webots_adapter::message_t<MAX_ROBOTS> blue_message;

        yellow_message.frame = frame;
        blue_message.frame = frame;

        for (size_t i = 0; i < robots_per_team; i++) {
            yellow_message.left_speed[i] = yellow_command.robot_command[i].left_speed;
            yellow_message.right_speed[i] = yellow_command.robot_command[i].right_speed;

            blue_message.left_speed[i] = blue_command.robot_command[i].left_speed;
            blue_message.right_speed[i] = blue_command.robot_command[i].right_speed;
        }

        yellow_team_emitter->send(&yellow_message, sizeof(yellow_message));
        blue_team_emitter->send(&blue_message, sizeof(blue_message));
    }

    return 0;
}

File: travesim-main/controllers/referee_controller/referee_controller.py
from controller import Supervisor
import json
import math
import socket
import struct
import sys


MAX_ROBOTS = 5
MESSAGE_FORMAT = "<I" + ("d" * MAX_ROBOTS) + ("d" * MAX_ROBOTS)


def parse_args():
    args = sys.argv[1:]
    return {
        "robots_per_team": int(args[0]) if len(args) > 0 else 3,
        "yellow_port": int(args[4]) if len(args) > 4 else 20012,
        "blue_port": int(args[6]) if len(args) > 6 else 20013,
        "vision_address": args[7] if len(args) > 7 else "224.0.0.1",
        "vision_port": int(args[8]) if len(args) > 8 else 10002,
    }


def yaw_from_orientation(orientation):
    # Webots returns a 3x3 orientation matrix. For planar robots, yaw is the
    # angle of the robot local X axis projected on the field XY plane.
    return math.atan2(orientation[3], orientation[0])


def node_state(node, robot_id=None):
    position = node.getPosition()
    velocity = node.getVelocity()
    state = {
        "x": position[0],
        "y": position[1],
        "z": position[2],
        "vx": velocity[0],
        "vy": velocity[1],
        "vz": velocity[2],
    }

    if robot_id is not None:
        state["robot_id"] = robot_id
        state["orientation"] = yaw_from_orientation(node.getOrientation())

    return state


def find_node_by_name(supervisor, name):
    children = supervisor.getRoot().getField("children")

    for index in range(children.getCount()):
        node = children.getMFNode(index)
        name_field = node.getField("name")

        if name_field is not None and name_field.getSFString() == name:
            return node

    return None


def command_bytes(frame, commands):
    left = [0.0] * MAX_ROBOTS
    right = [0.0] * MAX_ROBOTS

    for robot_id, speeds in commands.items():
        if 0 <= robot_id < MAX_ROBOTS:
            left[robot_id] = speeds[0]
            right[robot_id] = speeds[1]

    return struct.pack(MESSAGE_FORMAT, frame, *left, *right)


def read_udp_commands(sock):
    commands = {}

    while True:
        try:
            data, _ = sock.recvfrom(65535)
        except BlockingIOError:
            break

        try:
            payload = json.loads(data.decode("utf-8"))
        except (UnicodeDecodeError, json.JSONDecodeError):
            continue

        for item in payload.get("commands", []):
            robot_id = int(item.get("robot_id", 0))
            left = float(item.get("left_wheel", 0.0))
            right = float(item.get("right_wheel", 0.0))
            commands[robot_id] = (left, right)

    return commands


config = parse_args()
supervisor = Supervisor()
timestep = int(supervisor.getBasicTimeStep())

yellow_emitter = supervisor.getDevice("yellow_team")
blue_emitter = supervisor.getDevice("blue_team")
yellow_emitter.setChannel(0)
blue_emitter.setChannel(1)

yellow_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
yellow_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
yellow_socket.bind(("127.0.0.1", config["yellow_port"]))
yellow_socket.setblocking(False)

blue_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
blue_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
blue_socket.bind(("127.0.0.1", config["blue_port"]))
blue_socket.setblocking(False)

vision_socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)

ball = find_node_by_name(supervisor, "VssBall")
yellow_robots = [
    find_node_by_name(supervisor, f"YellowRobot{i}") for i in range(config["robots_per_team"])
]
blue_robots = [
    find_node_by_name(supervisor, f"BlueRobot{i}") for i in range(config["robots_per_team"])
]

if ball is None:
    raise RuntimeError("No se encontro el nodo VssBall")

if any(node is None for node in yellow_robots + blue_robots):
    raise RuntimeError("No se encontraron todos los robots por DEF/name")

frame = 0
last_yellow_commands = {}
last_blue_commands = {}

print("--- referee_controller.py iniciado ---")
print(f"Vision JSON: 127.0.0.1:{config['vision_port']}")
print(f"Yellow commands JSON: 127.0.0.1:{config['yellow_port']}")
print(f"Blue commands JSON: 127.0.0.1:{config['blue_port']}")

while supervisor.step(timestep) != -1:
    frame += 1

    new_yellow = read_udp_commands(yellow_socket)
    new_blue = read_udp_commands(blue_socket)

    if new_yellow:
        last_yellow_commands.update(new_yellow)
    if new_blue:
        last_blue_commands.update(new_blue)

    yellow_emitter.send(command_bytes(frame, last_yellow_commands))
    blue_emitter.send(command_bytes(frame, last_blue_commands))

    env = {
        "step": frame,
        "frame": {
            "ball": node_state(ball),
            "robots_yellow": [
                node_state(node, i) for i, node in enumerate(yellow_robots)
            ],
            "robots_blue": [
                node_state(node, i) for i, node in enumerate(blue_robots)
            ],
        },
        "field": {
            "width": 1.3,
            "length": 1.5,
            "goal_width": 0.4,
            "goal_depth": 0.1,
        },
    }

    vision_payload = json.dumps(env).encode("utf-8")

    vision_socket.sendto(
        vision_payload,
        ("127.0.0.1", config["vision_port"]),
    )
